module/x/gravity/keeper/invariants.go

```diff
@@ -0,0 +1,112 @@
+package keeper
+
+import (
+	"fmt"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	accType "github.com/cosmos/cosmos-sdk/x/auth/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/peggyjv/gravity-bridge/module/v2/x/gravity/types"
+)
+
+// TODO: Add any future invariants here
+// TODO: (see the sdk docs for more info https://docs.cosmos.network/master/building-modules/invariants.html)
+func AllInvariants(k Keeper) sdk.Invariant {
+	return func(ctx sdk.Context) (string, bool) {
+		return ModuleBalanceInvariant(k)(ctx)
+
+		// Example additional invariants
+		//  res, stop := FutureInvariant(k)(ctx)
+		//	if stop {
+		//		return res, stop
+		//	}
+		//
+		//	return AnotherFutureInvariant(k)(ctx)
+	}
+}
+
+// ModuleBalanceInvariant checks that the module account's balance is equal to the balance of unbatched transactions and unobserved batches
+// Note that the returned bool should be true if there is an error, e.g. an unexpected module balance
+func ModuleBalanceInvariant(k Keeper) sdk.Invariant {
+	return func(ctx sdk.Context) (string, bool) {
+		modAcc := accType.NewModuleAddress(types.ModuleName)
+		actualBals := k.bankKeeper.GetAllBalances(ctx, modAcc)
+		expectedBals := make(map[string]*sdk.Int, len(actualBals)) // Collect balances by contract
+		for _, v := range actualBals {
+			newInt := sdk.NewInt(0)
+			expectedBals[v.Denom] = &newInt
+		}
+		expectedBals = sumUnconfirmedBatchModuleBalances(ctx, k, expectedBals)
+		expectedBals = sumUnbatchedSendToEthereumsModuleBalances(ctx, k, expectedBals)
+
+		// Compare actual vs expected balances
+		for _, actual := range actualBals {
```

Comment:
should this also assert some metadata about the denoms? e.g. decimals if stored (I recall there was one PR adding that)?

Reply:
the decimals for the denom is used only for the ethereum contract metadata, but we are using only the base denom on chain. not sure what to check here